src/clj/puppetlabs/services/jruby/jruby_puppet_agents.clj

```diff
@@ -83,6 +83,7 @@
       (let [id (inc i)
             instance (jruby-core/borrow-from-pool (:pool old-pool))]
         (try
+          (.runScriptlet (:scripting-container instance) "Puppet::Server::Master.terminate")
```

Comment:
This is a nitpick, but I think that it might be better to add the 'terminate' method to the interface defined in `JRubyPuppet.java`, and then you can call that method directly on the `:jruby-puppet` object.  I'd prefer to try to minimize places where we're interacting directly with the `ScriptingContainer` as it is a less controlled code path.